Support structured logging

bells17 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

In this PR I have added JSON logging.
With this change, it is now possible to output the logs of the livenessprove container in JSON format.
Running the container with the `--logging-format=json` option will output the logs in JSON format.

In addition, I've modified the log messages based on the following guideline:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#remove-string-formatting-from-log-message

I’ve update the klog functions in use according to the guidelines provided below, and I've confirmed that they pass the [logcheck](https://github.com/kubernetes-sigs/logtools/tree/main/logcheck) tests:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#change-log-functions

```
$ logcheck cmd/livenessprobe/main.go
```

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes-csi/livenessprobe/issues/165

**Special notes for your reviewer**:

```
# The logs below were outputted because the program was running on MacOS.
$ go run cmd/livenessprobe/main.go --logging-format json
{"ts":1691826670746.745,"caller":"metrics/processstarttime.go:40","msg":"Could not get process start time, could not read \"/proc\": stat /proc: no such file or directory\n"}
{"ts":1691826670746.866,"caller":"metrics/processstarttime.go:40","msg":"Could not get process start time, could not read \"/proc\": stat /proc: no such file or directory\n"}
...

$ go run cmd/livenessprobe/main.go --logging-format text
E0812 16:51:25.016062   48838 processstarttime.go:40] Could not get process start time, could not read "/proc": stat /proc: no such file or directory
E0812 16:51:25.016189   48838 processstarttime.go:40] Could not get process start time, could not read "/proc": stat /proc: no such file or directory
...

$ go run cmd/livenessprobe/main.go
E0812 16:51:30.570105   48865 processstarttime.go:40] Could not get process start time, could not read "/proc": stat /proc: no such file or directory
E0812 16:51:30.570171   48865 processstarttime.go:40] Could not get process start time, could not read "/proc": stat /proc: no such file or directory
...

$ go run cmd/livenessprobe/main.go --logging-format unknown
E0812 17:13:34.539878   50062 main.go:144] "LoggingConfiguration is invalid" err="format: Invalid value: \"unknown\": Unsupported log format"
exit status 1
```

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Added support for structured logging (the log messages have been changed due to the activation of structured logging)
```

jsafrane commented:
/ok-to-test

jsafrane commented on cmd/livenessprobe/main.go:
```diff
@@ -119,16 +125,37 @@ func acquireConnection(ctx context.Context, metricsManager metrics.CSIMetricsMan
-	klog.InitFlags(nil)
+	var fs flag.FlagSet
+	klog.InitFlags(&fs)
+	fs.VisitAll(func(f *flag.Flag) {
+		switch f.Name {
+		case "logtostderr":
+			flag.CommandLine.Var(f.Value, f.Name, f.Usage)
+		}
+	})
```
Can you please add a comment why do we need this VisitAll?

bells17 replied:
Thank you for pointing it out. I've added a comment to the source code based on your feedback.

jsafrane commented on cmd/livenessprobe/main.go:
```diff
@@ -119,16 +125,37 @@ func acquireConnection(ctx context.Context, metricsManager metrics.CSIMetricsMan
+			flag.CommandLine.Var(f.Value, f.Name, f.Usage)
+		}
+	})
+	fg := featuregate.NewFeatureGate()
+	logsapi.AddFeatureGates(fg)
+	c := logsapi.NewLoggingConfiguration()
+	logsapi.AddGoFlags(c, flag.CommandLine)
+	logs.InitLogs()
 	flag.Set("logtostderr", "true")
 	flag.Parse()
+	if err := logsapi.ValidateAndApply(c, fg); err != nil {
+		klog.ErrorS(err, "LoggingConfiguration is invalid")
+		os.Exit(1)
+	}
+	if c.Format == logsapi.JSONLogFormat {
+		logger, _ := json.NewJSONLogger(c.Verbosity, os.Stderr, nil, nil)
+		klog.SetLogger(logger)
+		defer klog.ClearLogger()
+	}
```
This looks like quite a big chunk of code that could live in [csi-lib-utils](https://github.com/kubernetes-csi/csi-lib-utils), so all sidecars can benefit from it. BTW, is there anything in `component-base` that we could use to set up logging at a sidecar startup?

bells17 replied:
I apologize for the oversight. On double-checking, I realized that it wasn't necessary to manually reset the log format, so I've fixed the code.

> This looks like a fairly large chunk of code that could live in csi-lib-utils so that all sidecars can benefit from it.

Indeed, even after the changes, the code is about 20 lines long. It might be a good idea to move this implementation to a function within csi-lib-utils.

jsafrane commented on cmd/livenessprobe/main.go:
```diff
@@ -119,16 +124,40 @@ func acquireConnection(ctx context.Context, metricsManager metrics.CSIMetricsMan
 
 func main() {
-	klog.InitFlags(nil)
+	// We run `klog.InitFlags` to configure the `logtostderr` option.
+	// Due to a conflict with `logsapi.AddGoFlags`, we set only
+	// the `logtostderr` option in `flag.CommandLine` using the `fs` variable
+	// to avoid errors.
+	//
+	// For options set by klog and component-base logs, see the links below:
+	// https://github.com/kubernetes/component-base/blob/v0.28.0-rc.0/logs/api/v1/options.go#L337-L355
+	// https://github.com/kubernetes/klog/blob/v2.100.1/klog.go#L400-L424
+	var fs flag.FlagSet
+	klog.InitFlags(&fs)
+	fs.VisitAll(func(f *flag.Flag) {
+		switch f.Name {
+		case "logtostderr":
+			flag.CommandLine.Var(f.Value, f.Name, f.Usage)
+		}
+	})
```
I checked that `--logtostderr` is enabled by default. IMO it makes sense to remove `flag.Set("logtostderr", "true")`  and related code from this PR - it should do no harm.

We carried `flag.Set("logtostderr", "true")` since the sidecar first PR and nobody remembers why.

bells17 replied:
I've removed the code related to `--logtostderr`. Could you please review it again?

jsafrane reviewed (commented):
We discussed this PR in our sig-stroage triage meeting and we think that structured logging would be a great thing to have.

However, we would like to merge it *after* we release the CSI sidecars for Kubernetes 1.28 - it should be soon-ish.

Would you be interested in changing all the other sidecars? It would be great if all of them had the same logging cmdline options + capabilities in 1.29. You don't need to do all of them, we could find other volunteers. The cmdline changes are simple enough, but all the logs would need to be updated from `klog.Infof` to `klog.InfoS`.

bells17 commented:
>Would you be interested in changing all the other sidecars? It would be great if all of them had the same logging cmdline options + capabilities in 1.29.


I'd be happy to take on the challenge!

> The cmdline changes are simple enough, but all the logs would need to be updated from `klog.Infof` to `klog.InfoS`.

Understood.
I've already updated this Pull Request to use `klog.InfoS`, so I would appreciate it if you could review it again.

As for `klog.Fatal`, since there isn't a `klog.FatalS`, would it be better to change `klog.Fatal` to `klog.ErrorS`?
(`klog.Fatal` internally executes `os.Exit`, so it might not need to be changed.)

bells17 commented:
> As for klog.Fatal, since there isn't a klog.FatalS, would it be better to change klog.Fatal to klog.ErrorS?
(klog.Fatal internally executes os.Exit, so it might not need to be changed.)

Based on the guidelines specified at the URL below, I've updated the klog logging functions to align with those recommendations.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#change-log-functions

And I became aware of a tool called [logcheck](https://github.com/kubernetes-sigs/logtools/tree/main/logcheck) recently. So I've run logcheck command against the `cmd/livenessprobe/main.go` and confirmed that it passes.

```
$ logcheck cmd/livenessprobe/main.go
```

bells17 commented:
@jsafrane, I have updated the klog functions based on the guidelines.
I'd appreciate it if you could review again. Thank you.

jsafrane commented:
/lgtm
/approve
Thanks for the PR! It looks great.

/hold
We want to merge it _after_ our release for Kubernetes 1.28 (livenessprobe v2.11.0)

mauriciopoppe commented:
Echoing on https://github.com/kubernetes-csi/livenessprobe/pull/202#discussion_r1294671042, would it be possible to move some code to csi-lib-utils? 

I think you should squash your commits into 1 too.

mauriciopoppe commented on deployment/kubernetes/livenessprobe-sidecar.yaml:
```diff
@@ -50,4 +50,5 @@
     args:
     - --v=4
     - --csi-address=/csi/csi.sock
+    - --logging-format=json
```
Should the default be json or still text?. A concern about this line is that when a CSI Driver author looks at this line they might change the flag value and that might break things like log ingestion downstream if for example it's configured to read doing regex on text instead of json

bells17 replied:
It seems to me that changing the default output format to json might cause problems in the user's system, so perhaps the default output format should remain as text.
(Since the content of the message has also been changed for structured logging, I think an announcement about that is necessary.)

mauriciopoppe commented on cmd/livenessprobe/main.go:
```diff
@@ -119,16 +124,23 @@ func acquireConnection(ctx context.Context, metricsManager metrics.CSIMetricsMan
 func main() {
-	klog.InitFlags(nil)
-	flag.Set("logtostderr", "true")
+	fg := featuregate.NewFeatureGate()
```
considering the amount of code that needs to be shared maybe it's ok to have it in this repo and not in csi-lib-utils

mauriciopoppe reviewed (commented):
I'd suggest to squash a few commits that are related, maybe the changes in vendor/ can go to their own folder

bells17 commented:
@mauriciopoppe Thank you for reviewing. I've addressed the comments you made, so could you please review it again?

> I'd suggest to squash a few commits that are related, maybe the changes in vendor/ can go to their own folder

Is my understanding correct as follows?

- Squash the commits into one
- Separate the addition of the vendor directory into its own commit

I've made these changes, so I would appreciate it if you could confirm that everything is in order.

> Echoing on #202 (comment), would it be possible to move some code to csi-lib-utils?

I believe you are saying that there is no need to move this to csi-lib-utils  as below, so I haven't made any changes:

> considering the amount of code that needs to be shared maybe it's ok to have it in this repo and not in csi-lib-utils

If my understanding is incorrect, please let me know.

mauriciopoppe commented:
After reading the commits in more detail it looks like because of k8s.io/componentbase it pulled go-logr which is now in another commit. Initially I thought that these changes were independent but it looks like they should be part of a single commit, what do you think?

bells17 commented:
Thank you for reviewing.

> After reading the commits in more detail it looks like because of k8s.io/componentbase it pulled go-logr which is now in another commit. Initially I thought that these changes were independent but it looks like they should be part of a single commit, what do you think?

I indeed agree that the changes are dependent. On the other hand, the number of changed files in the vendor directory is quite large, so I feel that splitting the commits for readability would be a good idea.

> About the setup code I think moving it to csi-lib-utils would be nice but I feel there's the flag.Parse line that I wouldn't know where to put:
> ...
> Anyways I believe we should keep the setup code in this repo, they're just a few lines anyways

Understood. If I come up with a good idea while working on structured logging support for other CSI Sidecars, I'll propose it again.

bells17 commented:
@mauriciopoppe I apologize, but I thought it would be simpler to squash the commits, so I have done so and pushed. I kindly ask for your review again.

bells17 commented:
@jsafrane @mauriciopoppe Sorry, I missed a few spots that needed corrections, so I've added another commit. 
When you have a moment, could you please review it again? 
If everything looks good, please comment about this commit and I'll squash the commits and push again.

mauriciopoppe commented on README.md:
```diff
@@ -80,8 +80,9 @@ spec:
+Several klog arguments such as `-v <log level>` and `-logging-format <log format>` are supported.
 
-* All glog / klog arguments are supported, such as `-v <log level>` or `-alsologtostderr`.
+* Some klog arguments are supported, such as `-v <log level>` or `-logging-format <log format>`.
```
I think it should be `--v` and `--logging-format`. Also `--v` might be coming from klog but `--logging-format` is coming from component base

bells17 replied:
Thank you for reviewing it.
I fixed this description. Could you please review it again?

mauriciopoppe commented on README.md:
```diff
@@ -82,7 +82,7 @@ spec:
 Several klog arguments such as `-v <log level>` and `-logging-format <log format>` are supported.
 
-* Some klog arguments are supported, such as `-v <log level>` or `-logging-format <log format>`.
+* [Arguments set by the `k8s.io/component-base/logs` package for klog](https://github.com/kubernetes/component-base/blob/v0.28.0-rc.0/logs/api/v1/options.go#L337-L355) are supported, such as `-v <log level>` and `-logging-format <log format>`.
```
I believe we use `--` instead of `-`. Please update line 83 and 85.

bells17 replied:
Sorry, I removed the mistake on line 83. I've also fixed the description on line 85 to use `--`.

bells17 replied:
Thanks. I squashed the commits.

bells17 commented:
@mauriciopoppe Thanks. I squashed the commits.
https://github.com/kubernetes-csi/livenessprobe/pull/202#discussion_r1310460744

mauriciopoppe commented:
/lgtm

mauriciopoppe commented:
/lgtm

Waiting for the [hold lift]( https://github.com/kubernetes-csi/livenessprobe/pull/202#issuecomment-1689845846) to get merged

jsafrane commented:
/hold cancel
I've tagged v2.11, so let's merge this to master to have it in the next release.
